Improved category selection for recipes

## Current state

In recipe-editing mode the complete category needs to be typed in by hand. Typos result in newly created categories requiring manual fixing of recipes.

## Proposed state

When typing in the categories field the typed string is compared to all existing categories. Initially the comparison could be done character by character, later maybe with some fuzzy search. The detected category or a list of best results should be presented to the user who can select the correct one by clicking with the mouse or selecting with the arrow keys + return.

## Alternative

An alternative for the selection would be a dropdown list in which all existing categories are presented. The user can select from the list to fill the category field of the recipe.


## Additional Considerations

This suggestion could also be relevant for selecting ingredients, keywords, and tools. Maybe a generic approach which is reusable for those cases might be a good thing.

@sam-19 
I wanted to have a look at this, but I’m not too familiar with these vue component imports. I tried to import them as

```
import Multiselect from '@nextcloud/vue/dist/Components/Multiselect'
```

but when making the app I got some error messages like

```
ERROR in ./node_modules/@nextcloud/event-bus/dist/ProxyBus.js
Module not found: Error: Can't resolve 'semver/functions/major' in '/Users/sebastian/Projects/Nextcloud/Cookbook/nextcloud-docker-debug/volumes/custom_apps/cookbook/node_modules/@nextcloud/event-bus/dist'
 @ ./node_modules/@nextcloud/event-bus/dist/ProxyBus.js 10:36-69
 @ ./node_modules/@nextcloud/event-bus/dist/index.js
...
```

Do I need to register the components somewhere? Could you help me out here? :)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -31,6 +31,8 @@
   [#390](https://github.com/nextcloud/cookbook/pull/390) @christianlupus
 - Automatic deployment of new releases to the nextcloud app store
   [#433](https://github.com/nextcloud/cookbook/pull/433) @christianlupus
+- Category and keyword selection from list of existing ones in recipe editor
+  [#402](https://github.com/nextcloud/cookbook/pull/402/) @seyfeb
 
 ### Changed
 - Switch of project ownership to neextcloud organization in GitHub

--- a/src/components/EditMultiselect.vue
+++ b/src/components/EditMultiselect.vue
@@ -3,7 +3,6 @@
         <label>{{ fieldLabel }}</label>
         <Multiselect
             class="edit_ms"
-		    v-model="localValue"
             v-bind="$attrs"
             v-on="$listeners"
             />
@@ -16,35 +15,14 @@ export default {
     name: "EditMultiselect",
     components: {
         Multiselect
-    },
+    },   
     props: {
-		value: {
-			default() {
-				return []
-			},
-		},
         fieldLabel: String,
     },
     data () {
         return {
         }
-    },
-    computed: {
-        localValue: {
-			get() {
-				if (this.trackBy && this.options
-					&& typeof this.value !== 'object'
-					&& this.options[this.value]) {
-					return this.options[this.value]
-				}
-				return this.value
-			},
-			set(value) {
-				this.$emit('update:value', value)
-				this.$emit('change', value)
-			},
-		},
-    },
+    }
 }
 </script>
 

--- a/src/components/EditMultiselect.vue
+++ b/src/components/EditMultiselect.vue
@@ -1,0 +1,83 @@
+<template>
+    <fieldset>
+        <label>{{ fieldLabel }}</label>
+        <Multiselect
+            class="edit_ms"
+		    v-model="localValue"
+            v-bind="$attrs"
+            v-on="$listeners"
+            />
+    </fieldset>
+</template>
+
+<script>
+import Multiselect from '@nextcloud/vue/dist/Components/Multiselect'
+export default {
+    name: "EditMultiselect",
+    components: {
+        Multiselect
+    },
+    props: {
+		value: {
+			default() {
+				return []
+			},
+		},
+        fieldLabel: String,
+    },
+    data () {
+        return {
+        }
+    },
+    computed: {
+        localValue: {
+			get() {
+				if (this.trackBy && this.options
+					&& typeof this.value !== 'object'
+					&& this.options[this.value]) {
+					return this.options[this.value]
+				}
+				return this.value
+			},
+			set(value) {
+				this.$emit('update:value', value)
+				this.$emit('change', value)
+			},
+		},
+    },
+}
+</script>
+
+<style scoped>
+
+fieldset {
+    margin-bottom: 1em;
+    width: 100%;
+}
+
+fieldset > * {
+    margin: 0;
+    float: left;
+}
+    @media(max-width:1199px) { fieldset > label {
+        display: block;
+        float: none;
+    }}
+fieldset > label {
+    display: inline-block;
+    width: 10em;
+    line-height: 18px;
+    font-weight: bold;
+    word-spacing: initial;
+    vertical-align: top;
+}
+
+.edit_ms {
+    width: calc(100% - 11em + 10px);
+}
+
+    @media(max-width:1199px) { .edit_ms {
+        width: 100%;
+    }}
+
+</style>

--- a/src/components/EditMultiselect.vue
+++ b/src/components/EditMultiselect.vue
@@ -2,7 +2,7 @@
     <fieldset>
         <label>{{ fieldLabel }}</label>
         <Multiselect
-            class="edit_ms"
+            class="edit-multiselect"
             v-bind="$attrs"
             v-on="$listeners"
             />
@@ -50,12 +50,38 @@ fieldset > label {
     vertical-align: top;
 }
 
-.edit_ms {
+.edit-multiselect {
     width: calc(100% - 11em + 10px);
 }
 
-    @media(max-width:1199px) { .edit_ms {
+    @media(max-width:1199px) { .edit-multiselect {
         width: 100%;
     }}
+</style>
 
+<style>
+#app
+.edit-multiselect
+.multiselect__tags {
+    height: auto;
+    min-height: 34px;
+}
+
+#app
+.edit-multiselect
+.multiselect__tags
+.multiselect__tags-wrap {
+    display: flex;
+    flex-wrap: wrap;
+    padding-bottom: 0px;
+}
+
+#app
+.edit-multiselect
+.multiselect__tags
+.multiselect__tags-wrap
+.multiselect__tag {
+    flex-basis: 50px;
+    margin-bottom: 3px;
+} 
 </style>

--- a/src/components/RecipeEdit.vue
+++ b/src/components/RecipeEdit.vue
@@ -7,8 +7,8 @@
         <EditTimeField :fieldName="'prepTime'" :fieldLabel="t('cookbook', 'Preparation time')" />
         <EditTimeField :fieldName="'cookTime'" :fieldLabel="t('cookbook', 'Cooking time')" />
         <EditTimeField :fieldName="'totalTime'" :fieldLabel="t('cookbook', 'Total time')" />
-        <EditMultiselect :fieldLabel="t('cookbook', 'Category')" :placeholder="t('cookbook', 'Choose category')" v-model="recipe['recipeCategory']" :options="allCategories" :taggable="true" :multiple="false" @tag="addCategory" />
-        <EditMultiselect :fieldLabel="t('cookbook', 'Keywords')" :placeholder="t('cookbook', 'Choose keywords')" v-model="selectedKeywords" :options="allKeywords" :taggable="true" :multiple="true" :tagWidth="60" @tag="addKeyword" />
+        <EditMultiselect :fieldLabel="t('cookbook', 'Category')" :placeholder="t('cookbook', 'Choose category')" v-model="recipe['recipeCategory']" :options="allCategories" :taggable="true" :multiple="false" :loading="isFetchingCategories" @tag="addCategory" />
+        <EditMultiselect :fieldLabel="t('cookbook', 'Keywords')" :placeholder="t('cookbook', 'Choose keywords')" v-model="selectedKeywords" :options="allKeywords" :taggable="true" :multiple="true" :tagWidth="60" :loading="isFetchingKeywords" @tag="addKeyword" />
         <EditInputField :fieldName="'recipeYield'" :fieldType="'number'" :fieldLabel="t('cookbook', 'Servings')" />
         <EditInputGroup :fieldName="'tool'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Tools')"  v-bind:createFieldsOnNewlines="true" />
         <EditInputGroup :fieldName="'recipeIngredient'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Ingredients')" v-bind:createFieldsOnNewlines="true" />
@@ -60,6 +60,8 @@ export default {
             cookTime: [0, 0],
             totalTime: [0, 0],
             allCategories: [],
+            isFetchingCategories: true,
+            isFetchingKeywords: true,
             allKeywords: [],
             selectedKeywords: [],
         }
@@ -112,7 +114,7 @@ export default {
         /**
          * Fetch and display recipe categories
          */
-        fetchCategories: function() {
+        fetchCategories: function() {            
             $.get(this.$window.baseUrl + '/categories').done((json) => {
                 json = json || []
                 this.allCategories = []
@@ -123,6 +125,7 @@ export default {
                         )
                     }
                 }
+                this.isFetchingCategories = false
             })
             .fail((e) => {
                 alert(t('cookbook', 'Failed to fetch categories'))
@@ -147,6 +150,7 @@ export default {
                         }
                     }
                 }
+                this.isFetchingKeywords = false
             })
             .fail((e) => {
                 alert(t('cookbook', 'Failed to fetch keywords'))
@@ -262,6 +266,19 @@ export default {
                     this.totalTime = [timeComps[1], timeComps[2]]
                 }
                 this.selectedKeywords = this.recipe['keywords'].split(',')
+                
+                // fallback if fetching all keywords fails
+                this.selectedKeywords.forEach(kw => {
+                    if (!this.allKeywords.includes(kw)) {
+                        this.allKeywords.push(kw)
+                    }
+                })
+
+                // fallback if fetching all categories fails
+                if (!this.allCategories.includes(this.recipe['recipeCategory'])) {
+                    this.allCategories.push(this.recipe['recipeCategory'])
+                }
+
                 // Always set the active page last!
                 this.$store.dispatch('setPage', { page: 'edit' })
             } else {

--- a/src/components/RecipeEdit.vue
+++ b/src/components/RecipeEdit.vue
@@ -7,8 +7,8 @@
         <EditTimeField :fieldName="'prepTime'" :fieldLabel="t('cookbook', 'Preparation time')" />
         <EditTimeField :fieldName="'cookTime'" :fieldLabel="t('cookbook', 'Cooking time')" />
         <EditTimeField :fieldName="'totalTime'" :fieldLabel="t('cookbook', 'Total time')" />
-        <EditMultiselect :fieldLabel="t('cookbook', 'Category')" :placeholder="t('cookbook', 'Choose category')" v-model="recipe['recipeCategory']" :options="categories" :taggable="true" :multiple="false" @tag="addCategory" />
-        <EditInputField :fieldName="'keywords'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Keywords (comma separated)')" />
+        <EditMultiselect :fieldLabel="t('cookbook', 'Category')" :placeholder="t('cookbook', 'Choose category')" v-model="recipe['recipeCategory']" :options="allCategories" :taggable="true" :multiple="false" @tag="addCategory" />
+        <EditMultiselect :fieldLabel="t('cookbook', 'Keywords')" :placeholder="t('cookbook', 'Choose keywords')" v-model="selectedKeywords" :options="allKeywords" :taggable="true" :multiple="true" :tagWidth="60" @tag="addKeyword" />
         <EditInputField :fieldName="'recipeYield'" :fieldType="'number'" :fieldLabel="t('cookbook', 'Servings')" />
         <EditInputGroup :fieldName="'tool'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Tools')"  v-bind:createFieldsOnNewlines="true" />
         <EditInputGroup :fieldName="'recipeIngredient'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Ingredients')" v-bind:createFieldsOnNewlines="true" />
@@ -59,7 +59,9 @@ export default {
             prepTime: [0, 0],
             cookTime: [0, 0],
             totalTime: [0, 0],
-            categories: [],
+            allCategories: [],
+            allKeywords: [],
+            selectedKeywords: [],
         }
     },
     watch: {
@@ -78,14 +80,28 @@ export default {
             let mins = this.totalTime[1].toString().padStart(2, '0')
             this.recipe.totalTime = 'PT' + hours + 'H' + mins + 'M'
         },
+        selectedKeywords: {
+            deep: true,
+            handler() {
+                // convert keyword array to comma-separated string
+                this.recipe['keywords'] = this.selectedKeywords.join()
+            }
+        }
     },
     methods: {
         /** 
          * Add newly created category and set as selected.
          */
-        addCategory (newCategory) {            
-            this.categories.push(newCategory)            
+        addCategory (newCategory) {
+            this.allCategories.push(newCategory)
             this.recipe['recipeCategory'] = newCategory
+        },
+        /** 
+         * Add newly created keyword.
+         */
+        addKeyword (newKeyword) {
+            this.allKeywords.push(newKeyword)
+            this.selectedKeywords.push(newKeyword)
         },
         addEntry: function(field, index, content='') {
             this.recipe[field].splice(index, 0, content)
@@ -99,10 +115,10 @@ export default {
         fetchCategories: function() {
             $.get(this.$window.baseUrl + '/categories').done((json) => {
                 json = json || []
-                this.categories = []
+                this.allCategories = []
                 for (let i=0; i<json.length; i++) {
                     if (json[i].name != '*') {
-                        this.categories.push(
+                        this.allCategories.push(
                             json[i].name,
                         )
                     }
@@ -110,6 +126,30 @@ export default {
             })
             .fail((e) => {
                 alert(t('cookbook', 'Failed to fetch categories'))
+                if (e && e instanceof Error) {
+                    throw e
+                }
+            })
+        },
+        /**
+         * Fetch and display recipe keywords
+         */
+        fetchKeywords: function() {
+            $.ajax(this.$window.baseUrl + '/keywords').done((json) => {
+                json = json || []
+                if (json) {
+                    this.allKeywords = []
+                    for (let i=0; i<json.length; i++) {
+                        if (json[i].name != '*') {
+                            this.allKeywords.push(
+                                json[i].name,
+                            )
+                        }
+                    }
+                }
+            })
+            .fail((e) => {
+                alert(t('cookbook', 'Failed to fetch keywords'))
                 if (e && e instanceof Error) {
                     throw e
                 }
@@ -204,6 +244,7 @@ export default {
         },
         setup: function() {
             this.fetchCategories()
+            this.fetchKeywords()
             if (this.$route.params.id) {
                 // Load the recipe from store and make edits to a local copy first
                 this.recipe = { ...this.$store.state.recipe }
@@ -220,6 +261,7 @@ export default {
                 if (timeComps) {
                     this.totalTime = [timeComps[1], timeComps[2]]
                 }
+                this.selectedKeywords = this.recipe['keywords'].split(',')
                 // Always set the active page last!
                 this.$store.dispatch('setPage', { page: 'edit' })
             } else {

--- a/src/components/RecipeEdit.vue
+++ b/src/components/RecipeEdit.vue
@@ -7,7 +7,7 @@
         <EditTimeField :fieldName="'prepTime'" :fieldLabel="t('cookbook', 'Preparation time')" />
         <EditTimeField :fieldName="'cookTime'" :fieldLabel="t('cookbook', 'Cooking time')" />
         <EditTimeField :fieldName="'totalTime'" :fieldLabel="t('cookbook', 'Total time')" />
-        <EditInputField :fieldName="'recipeCategory'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Category')" />
+        <EditMultiselect :fieldLabel="t('cookbook', 'Category')" :placeholder="t('cookbook', 'Choose category')" v-model="recipe['recipeCategory']" :options="categories" :taggable="true" :multiple="false" @tag="addCategory" />
         <EditInputField :fieldName="'keywords'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Keywords (comma separated)')" />
         <EditInputField :fieldName="'recipeYield'" :fieldType="'number'" :fieldLabel="t('cookbook', 'Servings')" />
         <EditInputGroup :fieldName="'tool'" :fieldType="'text'" :fieldLabel="t('cookbook', 'Tools')"  v-bind:createFieldsOnNewlines="true" />
@@ -20,6 +20,7 @@
 import EditImageField from './EditImageField'
 import EditInputField from './EditInputField'
 import EditInputGroup from './EditInputGroup'
+import EditMultiselect from './EditMultiselect'
 import EditTimeField from './EditTimeField'
 
 export default {
@@ -28,6 +29,7 @@ export default {
         EditImageField,
         EditInputField,
         EditInputGroup,
+        EditMultiselect,
         EditTimeField,
     },
     props: ['id'],
@@ -57,6 +59,7 @@ export default {
             prepTime: [0, 0],
             cookTime: [0, 0],
             totalTime: [0, 0],
+            categories: [],
         }
     },
     watch: {
@@ -77,11 +80,40 @@ export default {
         },
     },
     methods: {
+        /** 
+         * Add newly created category and set as selected.
+         */
+        addCategory (newCategory) {            
+            this.categories.push(newCategory)            
+            this.recipe['recipeCategory'] = newCategory
+        },
         addEntry: function(field, index, content='') {
             this.recipe[field].splice(index, 0, content)
         },
         deleteEntry: function(field, index) {
             this.recipe[field].splice(index, 1)
+        },
+        /**
+         * Fetch and display recipe categories
+         */
+        fetchCategories: function() {
+            $.get(this.$window.baseUrl + '/categories').done((json) => {
+                json = json || []
+                this.categories = []
+                for (let i=0; i<json.length; i++) {
+                    if (json[i].name != '*') {
+                        this.categories.push(
+                            json[i].name,
+                        )
+                    }
+                }
+            })
+            .fail((e) => {
+                alert(t('cookbook', 'Failed to fetch categories'))
+                if (e && e instanceof Error) {
+                    throw e
+                }
+            })
         },
         loadRecipeData: function() {
             if (!this.$store.state.recipe) {
@@ -171,6 +203,7 @@ export default {
             }
         },
         setup: function() {
+            this.fetchCategories()
             if (this.$route.params.id) {
                 // Load the recipe from store and make edits to a local copy first
                 this.recipe = { ...this.$store.state.recipe }

